tasks/main.yml

```diff
@@ -75,15 +75,22 @@
       - "docker-compose-plugin"
     state: "{{ docker__state }}"
 
-- name: Install Python packages
-  ansible.builtin.pip:
-    name: >
-      {{ item.name }}{% if item.version | d() %}=={{ item.version }}{% endif %}
-    virtualenv: "{{ docker__pip_virtualenv }}"
-    virtualenv_python: "python3"
-    state: "{{ item.state | d('present') }}"
-  loop: "{{ docker__default_pip_packages + docker__pip_packages }}"
-  when: item.name | d()
+- block:
+    - name: Install Python packages
+      ansible.builtin.pip:
+        name: >
+          {{ item.name }}{% if item.version | d() %}=={{ item.version }}{% endif %}
+        virtualenv: "{{ docker__pip_virtualenv }}"
+        virtualenv_python: "python3"
+        state: "{{ item.state | d('present') }}"
+      loop: "{{ docker__default_pip_packages + docker__pip_packages }}"
+      when: item.name | d()
+  rescue:
+    - name: Handle upgrade case
+      ansible.builtin.file:
+        path: /usr/local/lib/docker/virtualenv
+        state: absent
+    - include_tasks: main.yml
```

Comment:
Is this going to re-run everything? I haven't used this pattern in Ansible yet.

Ideally I wonder if we can detect an upgrade occurred before the task runs so we can delete the file before this task runs.

Then we wouldn't need to rescue anything and the whole feature becomes maybe a single task to detect if it an upgrade happened and delete the file, else it does nothing.

Reply:
You're right, it did re-run everything. Just pushed a new version that is able to detect in advance and not do this.